lib/bolt/cli.rb

```diff
@@ -837,109 +849,82 @@ def initialize_project
-      result = resolver.resolve(
-        cache:                 nil,
-        ui:                    nil,
-        module_paths:          [],
-        allow_missing_modules: true
+      install_modules(
+        config.puppetfile,
+        config.puppetfile_config,
+        config.project.path + '.modules',
+        config.project.modules
       )
+    end
 
-      # Validate that the modules exist
-      missing_graph = result.specifications.select do |_name, spec|
-        spec.instance_of? PuppetfileResolver::Models::MissingModuleSpecification
-      end
-
-      if missing_graph.any?
-        titles = model.modules.each_with_object({}) do |mod, acc|
-          acc[mod.name] = mod.title
+    # Installs modules declared in the project configuration file.
+    #
+    def install_modules(puppetfile_path, config, moduledir, modules)
+      require 'bolt/puppetfile'
+      require 'bolt/puppetfile/installer'
+
+      puppetfile = Bolt::Puppetfile.new(modules)
+
+      # If the Puppetfile exists, check if it includes specs for each declared
+      # module, erroring if there are any missing. Otherwise, resolve the
+      # module dependencies and write a new Puppetfile. Users can forcibly
+      # overwrite an existing Puppetfile with the '--force' option.
+      if puppetfile_path.exist? && !options[:force]
+        outputter.print_message "Parsing existing Puppetfile at #{puppetfile_path}"
+        existing = Bolt::Puppetfile.parse(puppetfile_path)
+
+        unless existing.modules.superset? puppetfile.modules
+          missing_modules = puppetfile.modules - existing.modules
+
+          raise Bolt::Error.new(
+            "Puppetfile #{puppetfile_path} is missing specifications for modules: "\
+            "#{missing_modules.map(&:title).join(', ')}. This may not be a Puppetfile "\
+            "managed by Bolt. To forcibly overwrite the Puppetfile, run with the "\
+            "'--force' option.",
+            'bolt/missing-module-specs'
+          )
         end
-
-        names = titles.values_at(*missing_graph.keys)
-        plural = names.count == 1 ? '' : 's'
-
-        raise Bolt::ValidationError,
-              "Unknown module name#{plural} #{names.join(', ')}"
-      end
-
-      # Filter the dependency graph to only include module specifications
-      spec_graph = result.specifications.select do |_name, spec|
-        spec.instance_of? PuppetfileResolver::Models::ModuleSpecification
+      else
+        outputter.print_message "Resolving module dependencies, this may take a moment"
+        puppetfile.resolve
+        outputter.print_message "Writing Puppetfile at #{puppetfile_path}"
+        puppetfile.write(puppetfile_path, force: true)
       end
 
-      # Map specification models to a Puppetfile specification
-      spec_graph.values.map do |spec|
-        "mod '#{spec.owner}-#{spec.name}', '#{spec.version}'"
-      end
-    end
+      outputter.print_message "Syncing modules from #{puppetfile_path} to #{moduledir}"
```

Comment:
This is a little misleading, since if the list of modules in the project matches the list in Puppetfile it will do nothing (including not updating outdated module versions). I think we should only print this if we're actually installing or updating anything, and can probably skip the install step to save some cycles if the installed modules match the project list.

Reply:
Oo, the plot thickens! This *does* update versions if `--force` is provided, even if the sets are equal.